### PR TITLE
*: add a reference count for StmtCtx

### LINKED ISSUE #27725: fatal error: sync: unlock of unlocked mutex
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
fatal error: sync: unlock of unlocked mutex

goroutine 998 [running]:
runtime.throw(0x3b9f84a, 0x1e)
	/usr/local/go/src/runtime/panic.go:1117 +0x72 fp=0xc00dd31378 sp=0xc00dd31348 pc=0x1224532
sync.throw(0x3b9f84a, 0x1e)
	/usr/local/go/src/runtime/panic.go:1103 +0x35 fp=0xc00dd31398 sp=0xc00dd31378 pc=0x1259535
sync.(*Mutex).unlockSlow(0xc0052f3be8, 0xc0ffffffff)
	/usr/local/go/src/sync/mutex.go:196 +0xd8 fp=0xc00dd313c0 sp=0xc00dd31398 pc=0x126ab38
sync.(*Mutex).Unlock(0xc0052f3be8)
	/usr/local/go/src/sync/mutex.go:190 +0x48 fp=0xc00dd313e0 sp=0xc00dd313c0 pc=0x126aa48
github.com/pingcap/tidb/sessionctx/stmtctx.(*StatementContext).CopTasksDetails(0xc0052f3b98, 0xc0088fe120)
	/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/sessionctx/stmtctx/stmtctx.go:711 +0xbd6 fp=0xc00dd31778 sp=0xc00dd313e0 pc=0x1f1c0f6
github.com/pingcap/tidb/util/expensivequery.genLogFields(0xeae4e, 0xc000a767e0, 0x5df6e80, 0xc043ea79bb27c55a, 0x3960ccef0404)
	/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/util/expensivequery/expensivequery.go:117 +0x308 fp=0xc00dd31ff8 sp=0xc00dd31778 pc=0x26ef2
48
github.com/pingcap/tidb/util/expensivequery.(*memoryUsageAlarm).recordSQL.func2(0xc00dd322a0)
	/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/util/expensivequery/memory_usage_alarm.go:195 +0x385 fp=0xc00dd32170 sp=0xc00dd31ff8 pc=0x2
6f48e5
github.com/pingcap/tidb/util/expensivequery.(*memoryUsageAlarm).recordSQL(0xc00dd33f38, 0x40a1688, 0xc0014d95f0)
	/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/util/expensivequery/memory_usage_alarm.go:213 +0x7cb fp=0xc00dd32498 sp=0xc00dd32170 pc=0x2
6f330b
github.com/pingcap/tidb/util/expensivequery.(*memoryUsageAlarm).doRecord(0xc00dd33f38, 0x31e5b7000, 0x13221298, 0x40a1688, 0xc0014d95f0)
	/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/util/expensivequery/memory_usage_alarm.go:145 +0x5b8 fp=0xc00dd32608 sp=0xc00dd32498 pc=0x2
6f24f8
github.com/pingcap/tidb/util/expensivequery.(*memoryUsageAlarm).alarm4ExcessiveMemUsage(0xc00dd33f38, 0x40a1688, 0xc0014d95f0)
	/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/util/expensivequery/memory_usage_alarm.go:121 +0x1c5 fp=0xc00dd33e08 sp=0xc00dd32608 pc=0x2
6f1b65
github.com/pingcap/tidb/util/expensivequery.(*Handle).Run(0xc0014880c0)
	/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/util/expensivequery/expensivequery.go:82 +0x3e7 fp=0xc00dd33fd8 sp=0xc00dd33e08 pc=0x26eeb4
7
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc00dd33fe0 sp=0xc00dd33fd8 pc=0x125f7e1
created by main.createServer
	/home/jenkins/agent/workspace/build_tidb_multi_branch_master/go/src/github.com/pingcap/tidb/tidb-server/main.go:632 +0x1c7
```
[tidb_stderr.log](https://github.com/pingcap/tidb/files/7089416/tidb_stderr.log)

### 4. What is your TiDB version? (Required)

```
Release Version: v5.2.0-alpha-790-gd58d39e94
Edition: Community
Git Commit Hash: d58d39e9476f2503a1e8790f78a0d25272d0aabe
Git Branch: master
UTC Build Time: 2021-08-31 09:53:04
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close #27725 

Problem Summary:

### Check List

Tests 

- [x] Manual test (add detailed scripts or steps below)
Using commit: 17e6297aeecf264c8c75c89d9824e8be3f025812
run 
``` sql
select 1;
select sleep(3);
select 1;
select 1;
```

Before this commit, tidb-server will panic.
After this commit, tidb-server run successfully.

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by wshwsh12 — APPROVED** [review]

File: util/memoryusagealarm/memoryusagealarm.go [hunk 1]
```diff
@@ -294,6 +294,7 @@ func (record *memoryUsageAlarm) getTop10SqlInfo(cmp func(i, j *util.ProcessInfo)
 			}
 			buf.WriteString("\n")
 		}
+		totalCnt--
```
Comment (windtalker) [inline util/memoryusagealarm/memoryusagealarm.go:297]: Don't need to update `info` here?

File: util/util.go [hunk 2]
```diff
@@ -106,6 +106,11 @@ func Str2Int64Map(str string) map[int64]struct{} {
 
 // GenLogFields generate log fields.
 func GenLogFields(costTime time.Duration, info *ProcessInfo, needTruncateSQL bool) []zap.Field {
+	if !info.RefCountOfStmtCtx.TryIncrease() {
```
Comment (windtalker) [inline util/util.go:109]: So other threads can only access `StmtCtx` by this method?

**Comment by hawkingrei** [comment]
@XuHuaiyu  PTAL

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4aa405]

goroutine 130 [running]:
testing.tRunner.func1.2({0x1ac2d80, 0x3047a00})
	GOROOT/src/testing/testing.go:1396 +0x372
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1399 +0x5f0
panic({0x1ac2d80, 0x3047a00})
	GOROOT/src/runtime/panic.go:890 +0x262
sync/atomic.LoadInt32()
	src/runtime/race_amd64.s:202 +0xc
github.com/pingcap/tidb/sessionctx/stmtctx.(*ReferenceCount).TryIncrease(0x0?)
	sessionctx/stmtctx/stmtctx.go:118 +0x45
github.com/pingcap/tidb/util.GenLogFields(0x0, 0xc0001a37a0, 0x0)
	util/util.go:109 +0xa5
github.com/pingcap/tidb/util/memoryusagealarm.(*memoryUsageAlarm).getTop10SqlInfo(0xc000611cd0, 0x0?, {0xc00059a798?, 0x3, 0x3})
	util/memoryusagealarm/memoryusagealarm.go:274 +0x2d4
github.com/pingcap/tidb/util/memoryusagealarm.(*memoryUsageAlarm).getTop10SqlInfoByMemoryUsage(0xc00080bb70?, {0xc00059a798, 0x3, 0x3})
	util/memoryusagealarm/memoryusagealarm.go:304 +0x67
github.com/pingcap/tidb/util/memoryusagealarm.TestGetTop10Sql(0x0?)
	util/memoryusagealarm/memoryusagealarm_test.go:87 +0x205
testing.tRunner(0xc000603040, 0x1d17040)
	GOROOT/src/testing/testing.go:1446 +0x217
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:1493 +0x75e
```

File: util/util.go (hunk 2 above)
Comment (XuHuaiyu) [reply util/util.go:109]: I'm not sure. But it seems we only encounter problems here.😅

**Review by windtalker — APPROVED** [review]
LGTM

**Review by qw4990 — APPROVED** [review]

**Pull request merged**